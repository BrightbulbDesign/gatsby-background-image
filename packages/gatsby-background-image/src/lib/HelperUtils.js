import filterInvalidDOMProps from 'filter-invalid-dom-props'
import {
  groupByMedia,
  hasArtDirectionFixedArray,
  hasArtDirectionFluidArray,
} from './MediaUtils'

/**
 * Are we in the browser?
 *
 * @return {boolean}
 */
export const isBrowser = () => typeof window !== 'undefined'

/**
 * Tests a given value on being a string.
 *
 * @param value *   Value to test
 * @return {boolean}
 */
export const isString = value =>
  Object.prototype.toString.call(value) === '[object String]'

/**
 * Strip BackgroundImage propTypes from remaining props to be passed to <Tag />
 *
 * @param props
 * @return {Object}
 */
export const stripRemainingProps = props => filterInvalidDOMProps(props)

/**
 * Handle legacy names for image queries
 *
 * @param props
 * @return {Object}
 */
export const convertProps = props => {
  const convertedProps = { ...props }
  const { resolutions, sizes, classId, fixed, fluid } = convertedProps

  if (resolutions) {
    convertedProps.fixed = resolutions
    delete convertedProps.resolutions
  }
  if (sizes) {
    convertedProps.fluid = sizes
    delete convertedProps.sizes
  }

  if (classId) {
    logDeprecationNotice(
      `classId`,
      `gatsby-background-image should provide unique classes automatically. Open an Issue should you still need this property.`
    )
  }

  // if (fluid && !hasImageArray(props)) {
  //   convertedProps.fluid = [].concat(fluid)
  // }
  //
  // if (fixed && !hasImageArray(props)) {
  //   convertedProps.fixed = [].concat(fixed)
  // }

  // convert fluid & fixed to arrays so we only have to work with arrays
  if (fluid && hasArtDirectionFluidArray(props)) {
    convertedProps.fluid = groupByMedia(convertedProps.fluid)
  }
  if (fixed && hasArtDirectionFixedArray(props)) {
    convertedProps.fixed = groupByMedia(convertedProps.fixed)
  }

  return convertedProps
}

/**
 * Checks if fluid or fixed are image arrays.
 *
 * @param props   object   The props to check for images.
 * @return {boolean}
 */
export const hasImageArray = props =>
  (props.fluid && Array.isArray(props.fluid)) ||
  (props.fixed && Array.isArray(props.fixed))

/**
 * Converts CSS kebab-case strings to camel-cased js style rules.
 *
 * @param str   string    Rule to transform
 * @return {boolean|string}
 */
export const toCamelCase = str =>
  isString(str) &&
  str
    .toLowerCase()
    .replace(/(?:^\w|-|[A-Z]|\b\w)/g, (letter, index) =>
      index === 0 ? letter.toLowerCase() : letter.toUpperCase()
    )
    .replace(/\s|\W+/g, '')

/**
 * Converts camel-cased js style rules to CSS kebab-case strings.
 *
 * @param str string    Rule to transform
 * @return {boolean|string}
 */
export const toKebabCase = str =>
  isString(str) &&
  str
    .replace(/\s|\W+/g, '')
    .replace(/[A-Z]/g, match => `-${match.toLowerCase()}`)

/**
 * Splits a given string (e.g. from classname) to an array.
 *
 * @param str string|array  String to split or return as array
 * @param delimiter string  Delimiter on which to split str
 * @return {array|boolean}  Returns (split) string as array, false on failure
 */
export const stringToArray = (str, delimiter = ` `) => {
  if (str instanceof Array) {
    return str
  }
  if (isString(str)) {
    if (str.includes(delimiter)) {
      return str.split(delimiter)
    }
    return [str]
  }
  return false
}

/**
 * Hashes a String to a 32bit integer with the simple Java 8 hashCode() func.
 *
 * @param str   string    String to hash.
 * @return {number}
 */
export const hashString = str =>
  isString(str) &&
  [].reduce.call(
    str,
    (hash, item) => {
      hash = (hash << 5) - hash + item.charCodeAt(0)
      return hash | 0
    },
    0
  )

/**
 * As the name says, it filters out empty strings from an array and joins it.
 *
 * @param arrayToJoin   array   Array to join after filtering.
 * @return {string}
 */
export const filteredJoin = arrayToJoin =>
  arrayToJoin.filter(item => item !== ``).join()

/**
 * Combines two arrays while keeping fromArrays indexes & values.
 *
 * @param fromArray   array   Array the values shall be taken from.
 * @param toArray     array   Array to copy values into.
 * @return {array}
 */
export const combineArray = (fromArray, toArray) => {
  // Fallback for singular images.
  if (!Array.isArray(fromArray)) {
    return [fromArray]
  }
  return fromArray.map((item, index) => item || toArray[index])
}

/**
 * Logs a warning if deprecated props where used.
 *
 * @param prop
 * @param notice
 */
export const logDeprecationNotice = (prop, notice) => {
  if (process.env.NODE_ENV === `production`) {
    return
  }

  console.log(
    `
    The "${prop}" prop is now deprecated and will be removed in the next major version
    of "gatsby-background-image".
    `
  )

  if (notice) {
    console.log(notice)
  }
}
