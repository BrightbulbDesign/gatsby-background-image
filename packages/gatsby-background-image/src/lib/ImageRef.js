import {
  convertProps,
  getCurrentSrcData,
  hasArtDirectionArray,
  hasImageArray,
  isBrowser,
} from './HelperUtils'
import {
  createArtDirectionSources,
  getCurrentFromData,
  hasPictureElement,
} from './ImageUtils'

/**
 * Creates an image reference to be activated on critical or visibility.
 * @param props
 * @param onLoad
 * @return {HTMLImageElement|null|Array}
 */
export const createPictureRef = (props, onLoad) => {
  const convertedProps = convertProps(props)

  if (
    isBrowser() &&
    (typeof convertedProps.fluid !== `undefined` ||
      typeof convertedProps.fixed !== `undefined`)
  ) {
    if (
      hasImageArray(convertedProps) &&
      !hasArtDirectionArray(convertedProps)
    ) {
      return createMultiplePictureRefs(props, onLoad)
    }
    const img = new Image()

    img.onload = () => onLoad()

    if (!img.complete && typeof convertedProps.onLoad === `function`) {
      img.addEventListener('load', convertedProps.onLoad)
    }
    if (typeof convertedProps.onError === `function`) {
      img.addEventListener('error', convertedProps.onError)
    }
    if (convertedProps.crossOrigin) {
      img.crossOrigin = convertedProps.crossOrigin
    }

    // Only directly activate the image if critical (preload).
    if (convertedProps.critical || convertedProps.isVisible) {
      return activatePictureRef(img, convertedProps)
    }

    return img
  }
  return null
}

/**
 * Creates multiple image references. Internal function.
 *
 * @param props   object    Component Props (with fluid or fixed as array).
 * @param onLoad  function  Callback for load handling.
 */
export const createMultiplePictureRefs = (props, onLoad) => {
  const convertedProps = convertProps(props)

  // Extract Image Array.
  const imageStack = convertedProps.fluid || convertedProps.fixed
  return imageStack.map(imageData => {
    if (convertedProps.fluid) {
      return createPictureRef({ ...convertedProps, fluid: imageData }, onLoad)
    }
    return createPictureRef({ ...convertedProps, fixed: imageData }, onLoad)
  })
}

/**
 * Creates a picture element for the browser to decide which image to load.
 *
 * @param imageRef
 * @param props
 * @param selfRef
 * @return {null|Array|*}
 */
export const activatePictureRef = (imageRef, props, selfRef = null) => {
  const convertedProps = convertProps(props)
  if (
    isBrowser() &&
    (typeof convertedProps.fluid !== `undefined` ||
      typeof convertedProps.fixed !== `undefined`)
  ) {
    if (
      hasImageArray(convertedProps) &&
      !hasArtDirectionArray(convertedProps)
    ) {
      return activateMultiplePictureRefs(imageRef, props, selfRef)
    } else {
      const imageData = hasArtDirectionArray(convertedProps)
        ? getCurrentSrcData(convertedProps)
        : getCurrentFromData(convertedProps)

      // Prevent adding HTMLPictureElement if it isn't supported (e.g. IE11),
      // but don't prevent it during SSR.
      let removableElement = null
      if (hasPictureElement()) {
        const pic = document.createElement('picture')
        if (selfRef) {
          // Set original component's style.
          pic.width = imageRef.width = selfRef.offsetWidth
          pic.height = imageRef.height = selfRef.offsetHeight
        }
        // TODO: check why only the 1400 image gets loaded & single / stacked images don't!
        if (hasArtDirectionArray(convertedProps)) {
          const sources = createArtDirectionSources(convertedProps)
          sources.forEach(currentSource => pic.appendChild(currentSource))
        } else if (imageData.srcSetWebp) {
          const sourcesWebP = document.createElement('source')
          sourcesWebP.type = `image/webp`
          sourcesWebP.srcset = imageData.srcSetWebp
          sourcesWebP.sizes = imageData.sizes
          if (imageData.media) {
            sourcesWebP.media = imageData.media
          }
          pic.appendChild(sourcesWebP)
        }
        pic.appendChild(imageRef)
        removableElement = pic
        // document.body.appendChild(removableElement)
      } else {
        if (selfRef) {
          imageRef.width = selfRef.offsetWidth
          imageRef.height = selfRef.offsetHeight
        }
        removableElement = imageRef
        // document.body.appendChild(removableElement)
      }

      imageRef.srcset = imageData.srcSet ? imageData.srcSet : ``
      imageRef.src = imageData.src ? imageData.src : ``
      if (imageData.media) {
        imageRef.media = imageData.media
      }

      // document.body.removeChild(removableElement)

      return imageRef
    }
  }
  return null
}

/**
 * Creates multiple picture elements.
 *
 * @param imageRefs
 * @param props
 * @param selfRef
 * @return {Array||null}
 */
export const activateMultiplePictureRefs = (imageRefs = [], props, selfRef) => {
  const convertedProps = convertProps(props)

  // Extract Image Array.
  return imageRefs.map((imageRef, index) => {
    if (convertedProps.fluid) {
      return activatePictureRef(
        imageRef,
        {
          ...convertedProps,
          fluid: convertedProps.fluid[index],
        },
        selfRef
      )
    }
    return activatePictureRef(
      imageRef,
      {
        ...convertedProps,
        fixed: convertedProps.fixed[index],
      },
      selfRef
    )
  })
}
