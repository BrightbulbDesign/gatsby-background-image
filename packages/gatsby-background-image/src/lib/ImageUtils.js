import {
  filteredJoin,
  hasArtDirectionArray,
  isString,
  getCurrentSrcData,
} from './HelperUtils'

/**
 * Returns the availability of the HTMLPictureElement unless in SSR mode.
 *
 * @return {boolean}
 */
export const hasPictureElement = () =>
  typeof HTMLPictureElement !== `undefined` || typeof window === `undefined`

/**
 * Creates a source Array from media objects.
 *
 * @param fluid
 * @param fixed
 * @return {*}
 */
export const createArtDirectionSources = ({ fluid, fixed }) => {
  const currentSource = fluid || fixed
  return currentSource.map(image => {
    const source = document.createElement('source')
    source.srcset = image.srcSet
    source.sizes = image.sizes
    if (image.srcSetWebp) {
      source.type = `image/webp`
      source.srcset = image.srcSetWebp
    }
    if (image.media) {
      source.media = image.media
    }
    return source
  })
}

/**
 * Extracts a value from an imageRef, image object or an array of them.
 *
 * @param data        HTMLImageElement||object||Array   Data to extract from.
 * @param propName    string    Property to extract.
 * @param addUrl      boolean   Should returned strings be encased in `url()`?
 * @param returnArray boolean   Switches between returning an array and a string.
 * @param checkLoaded boolean   Turns checking for imageLoaded() on and off.
 * @return {string||array}
 */
export const getCurrentFromData = ({
  data,
  propName,
  addUrl = true,
  returnArray = false,
  checkLoaded = true,
}) => {
  if (!data || !propName) return ``
  // Handle tracedSVG with "special care".
  const tracedSVG = propName === `tracedSVG`
  if (Array.isArray(data) && !hasArtDirectionArray({ fluid: data })) {
    // Filter out all elements not having the propName and return remaining.
    const imageString = data
      // .filter(dataElement => {
      //   return propName in dataElement && dataElement[propName]
      // })
      .map(dataElement => {
        // If `currentSrc` or `src` is needed, check image load completion first.
        if (propName === `currentSrc` || propName === 'src') {
          return checkLoaded
            ? (imageLoaded(dataElement) && dataElement[propName]) || ``
            : dataElement[propName]
        }
        // Check if CSS strings should be parsed.
        if (propName === `CSS_STRING` && isString(dataElement)) {
          return dataElement
        }
        return dataElement[propName] || ``
      })
    // Encapsulate in URL string and return.
    return getUrlString({
      imageString,
      tracedSVG,
      addUrl,
      returnArray,
    })
  }
  if (
    hasArtDirectionArray({ fluid: data }) &&
    (propName === `currentSrc` || propName === 'src')
  ) {
    const currentData = getCurrentSrcData({ fluid: data })
    return propName in currentData
      ? getUrlString({ imageString: currentData[propName], tracedSVG, addUrl })
      : ``
  }
  // If `currentSrc` or `src` is needed, check image load completion first.
  if ((propName === `currentSrc` || propName === 'src') && propName in data) {
    return getUrlString({
      imageString: checkLoaded
        ? (imageLoaded(data) && data[propName]) || ``
        : data[propName],
      addUrl,
    })
  }
  return propName in data
    ? getUrlString({ imageString: data[propName], tracedSVG, addUrl })
    : ``
}

/**
 * Encapsulates an imageString with a url if needed.
 *
 * @param imageString   string    String to encapsulate.
 * @param tracedSVG     boolean   Special care for SVGs.
 * @param addUrl        boolean   If the string should be encapsulated or not.
 * @param returnArray   boolean   Return concatenated string or Array.
 * @param hasImageUrls  boolean   Force return of quoted string(s) for url().
 * @return {string||array}
 */
export const getUrlString = ({
  imageString,
  tracedSVG = false,
  addUrl = true,
  returnArray = false,
  hasImageUrls = false,
}) => {
  if (Array.isArray(imageString)) {
    const stringArray = imageString.map(currentString => {
      if (currentString) {
        const base64 = currentString.indexOf(`base64`) !== -1
        const imageUrl = hasImageUrls || currentString.substr(0, 4) === `http`
        const currentReturnString =
          currentString && tracedSVG
            ? `"${currentString}"`
            : currentString && !base64 && !tracedSVG && imageUrl
            ? `'${currentString}'`
            : currentString
        return addUrl && currentString
          ? `url(${currentReturnString})`
          : currentReturnString
      }
      return currentString
    })
    return returnArray ? stringArray : filteredJoin(stringArray)
  }
  const base64 = imageString.indexOf(`base64`) !== -1
  const imageUrl = hasImageUrls || imageString.substr(0, 4) === `http`
  const returnString =
    imageString && tracedSVG
      ? `"${imageString}"`
      : imageString && !base64 && !tracedSVG && imageUrl
      ? `'${imageString}'`
      : imageString
  return imageString ? (addUrl ? `url(${returnString})` : returnString) : ``
}

/**
 * Checks if any image props have changed.
 *
 * @param props
 * @param prevProps
 * @return {*}
 */
export const imagePropsChanged = (props, prevProps) =>
  // Do we have different image types?
  (props.fluid && !prevProps.fluid) ||
  (props.fixed && !prevProps.fixed) ||
  imageArrayPropsChanged(props, prevProps) ||
  // Are single image sources different?
  (props.fluid && prevProps.fluid && props.fluid.src !== prevProps.fluid.src) ||
  (props.fixed && prevProps.fixed && props.fixed.src !== prevProps.fixed.src)

/**
 * Decides if two given props with array images differ.
 *
 * @param props
 * @param prevProps
 * @return {boolean}
 */
export const imageArrayPropsChanged = (props, prevProps) => {
  const isPropsFluidArray = Array.isArray(props.fluid)
  const isPrevPropsFluidArray = Array.isArray(prevProps.fluid)
  const isPropsFixedArray = Array.isArray(props.fixed)
  const isPrevPropsFixedArray = Array.isArray(prevProps.fixed)

  if (
    // Did the props change to a single image?
    (isPropsFluidArray && !isPrevPropsFluidArray) ||
    (isPropsFixedArray && !isPrevPropsFixedArray) ||
    // Did the props change to an Array?
    (!isPropsFluidArray && isPrevPropsFluidArray) ||
    (!isPropsFixedArray && isPrevPropsFixedArray)
  ) {
    return true
  }
  // Are the lengths or sources in the Arrays different?
  if (isPropsFluidArray && isPrevPropsFluidArray) {
    if (props.fluid.length === prevProps.fluid.length) {
      // Check for individual image or CSS string changes.
      return props.fluid.some(
        (image, index) => image.src !== prevProps.fluid[index].src
      )
    }
    return true
  } else if (isPropsFixedArray && isPrevPropsFixedArray) {
    if (props.fixed.length === prevProps.fixed.length) {
      // Check for individual image or CSS string changes.
      return props.fixed.some(
        (image, index) => image.src !== prevProps.fixed[index].src
      )
    }
    return true
  }
}

/**
 * Creates an array with a transparent dummy pixel for background-* properties.
 *
 * @param length
 * @return {any[]}
 */
export const createDummyImageArray = length => {
  const DUMMY_IMG = `data:image/gif;base64,R0lGODlhAQABAAAAACH5BAEKAAEALAAAAAABAAEAAAICTAEAOw==`
  const dummyImageURI = getUrlString({ imageString: DUMMY_IMG })
  return Array(length).fill(dummyImageURI)
}

/**
 * Checks if an image (array) reference is existing and tests for complete.
 *
 * @param imageRef    HTMLImageElement||array   Image reference(s).
 * @return {boolean}
 */
export const imageReferenceCompleted = imageRef =>
  imageRef
    ? Array.isArray(imageRef)
      ? imageRef.every(singleImageRef => imageLoaded(singleImageRef))
      : imageLoaded(imageRef)
    : false

/**
 * Checks if an image really was fully loaded.
 *
 * @param imageRef  HTMLImageElement  Reference to an image.
 * @return {boolean}
 */
export const imageLoaded = imageRef =>
  imageRef
    ? imageRef.complete &&
      imageRef.naturalWidth !== 0 &&
      imageRef.naturalHeight !== 0
    : false
