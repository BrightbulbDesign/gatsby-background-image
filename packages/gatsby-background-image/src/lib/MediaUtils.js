import smq from 'sort-media-queries'
import { isBrowser } from './SimpleUtils'

/**
 * Return an array ordered by elements having a media prop, does not use
 * native sort, as a stable sort is not guaranteed by all browsers/versions
 *
 * @param imageVariants   array   The art-directed images.-
 */
export const groupByMedia = imageVariants => {
  const without = []
  const sortedVariants = smq(imageVariants, 'media')

  sortedVariants.forEach(variant => !variant.media && without.push(variant))

  if (without.length > 1 && process.env.NODE_ENV !== `production`) {
    console.warn(
      `We've found ${without.length} sources without a media property. They might be ignored by the browser, see: https://www.gatsbyjs.org/packages/gatsby-image/#art-directing-multiple-images`
    )
  }

  return sortedVariants
}

/**
 * Creates a source Array from media objects.
 *
 * @param fluid
 * @param fixed
 * @return {*}
 */
export const createArtDirectionSources = ({ fluid, fixed }) => {
  const currentSource = fluid || fixed
  return currentSource.reduce((sources, image) => {
    if (!image.media) {
      return sources
    }
    const source = document.createElement('source')
    source.srcset = image.srcSet
    source.media = image.media
    if (image.sizes) {
      source.sizes = image.sizes
    }
    if (image.srcSetWebp) {
      source.type = `image/webp`
      source.srcSetWebp = image.srcSetWebp
    }
    source.media = image.media
    sources.push(source)
    return sources
  }, [])
}

/**
 * Checks if fluid or fixed are art-direction arrays.
 *
 * @param props   object   The props to check for images.
 * @param prop    string   Check for fluid or fixed.
 * @return {boolean}
 */
export const hasArtDirectionSupport = (props, prop) =>
  props[prop] &&
  Array.isArray(props[prop]) &&
  props[prop].some(image => typeof image.media !== 'undefined')

/**
 * Checks for fluid or fixed Art direction support.
 * @param props
 * @return {boolean}
 */
export const hasArtDirectionArray = props =>
  hasArtDirectionSupport(props, 'fluid') ||
  hasArtDirectionSupport(props, 'fixed')

/**
 * Tries to detect if a media query matches the current viewport.
 *
 * @param media   string  A media query string.
 * @return {boolean}
 */
export const matchesMedia = ({ media }) => {
  return media
    ? isBrowser() &&
        typeof window.matchMedia !== `undefined` &&
        !!window.matchMedia(media).matches
    : false
}
