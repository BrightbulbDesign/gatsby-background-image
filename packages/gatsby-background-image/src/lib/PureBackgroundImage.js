import React from 'react'
import PropTypes from 'prop-types'

class PureBackgroundImage extends React.PureComponent {
  render() {
    const {
      Tag,
      currentClassNames,
      currentStyles,
      handleRef,
      componentKey,
      remainingProps,
      pseudoStyles,
      hasNoScript,
      noScriptPseudoStyles,
      children,
    } = this.props
    return (
      <Tag
        className={currentClassNames}
        style={currentStyles}
        ref={handleRef}
        key={componentKey}
        {...remainingProps}
      >
        {/* Create style element to transition between pseudo-elements. */}
        {/* TODO: try to memoize this */}
        <style
          dangerouslySetInnerHTML={{
            __html: pseudoStyles,
          }}
        />
        {/* Set the original image(s) during SSR & if JS is disabled */}
        {hasNoScript && (
          <noscript>
            <style
              dangerouslySetInnerHTML={{
                __html: noScriptPseudoStyles,
              }}
            />
          </noscript>
        )}
        {children}
      </Tag>
    )
  }
}

PureBackgroundImage.propTypes = {
  Tag: PropTypes.oneOfType([PropTypes.string, PropTypes.func]).isRequired,
  currentClassNames: PropTypes.string.isRequired,
  currentStyles: PropTypes.oneOfType([PropTypes.object, PropTypes.array])
    .isRequired,
  pseudoStyles: PropTypes.string.isRequired,
  handleRef: PropTypes.func.isRequired,
  componentKey: PropTypes.string.isRequired,
  hasNoScript: PropTypes.bool.isRequired,
  noScriptPseudoStyles: PropTypes.string.isRequired,
  // eslint-disable-next-line react/forbid-prop-types
  remainingProps: PropTypes.any,
}

export default PureBackgroundImage
