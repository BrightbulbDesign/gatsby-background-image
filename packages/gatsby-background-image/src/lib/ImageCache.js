import {
  convertProps,
  hasImageArray,
} from './HelperUtils'
import { hasArtDirectionArray } from './MediaUtils'
import { getImageSrcKey } from './ImageUtils'

const imageCache = Object.create({})
/**
 * Cache if we've seen an image before so we don't both with
 * lazy-loading & fading in on subsequent mounts.
 *
 * @param props
 * @return {boolean}
 */
export const inImageCache = props => {
  const convertedProps = convertProps(props)
  if (hasImageArray(convertedProps) && !hasArtDirectionArray(convertedProps)) {
    return allInImageCache(props)
  }

  // Find src
  const src = getImageSrcKey(convertedProps)
  return imageCache[src] || false
}

/**
 * Processes an array of cached images for inImageCache.
 *
 * @param props  object    Component Props (with fluid or fixed as array).
 * @return {*|boolean}
 */
export const allInImageCache = props => {
  const convertedProps = convertProps(props)
  // Extract Image Array.
  const imageStack = convertedProps.fluid || convertedProps.fixed
  // Only return true if every image is in cache.
  return imageStack.every(imageData => {
    if (convertedProps.fluid) {
      return inImageCache({ fluid: imageData })
    }
    return inImageCache({ fixed: imageData })
  })
}

/**
 * Adds an Image to imageCache.
 *
 * @param props
 */
export const activateCacheForImage = props => {
  const convertedProps = convertProps(props)
  if (hasImageArray(convertedProps)) {
    return activateCacheForMultipleImages(props)
  }
  // Find src
  const src = getImageSrcKey(convertedProps)
  if (src) {
    imageCache[src] = true
  }
}

/**
 * Activates the Cache for multiple Images.
 *
 * @param props
 */
export const activateCacheForMultipleImages = props => {
  const convertedProps = convertProps(props)
  // Extract Image Array.
  const imageStack = convertedProps.fluid || convertedProps.fixed
  imageStack.forEach(imageData => {
    if (convertedProps.fluid) {
      activateCacheForImage({ fluid: imageData })
    } else {
      activateCacheForImage({ fixed: imageData })
    }
  })
}

/**
 * Resets the image cache (especially important for reliable tests).
 */
export const resetImageCache = () => {
  for (const prop in imageCache) delete imageCache[prop]
}
