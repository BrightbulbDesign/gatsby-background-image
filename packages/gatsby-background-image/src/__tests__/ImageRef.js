import {
  fixedArrayMock,
  fixedShapeMock,
  fluidArrayMock,
  fluidShapeMock,
  mockArtDirectionStackFluid,
} from './mocks/Various.mock'
import { activatePictureRef, createPictureRef } from '../lib/ImageRef'

describe(`createPictureRef()`, () => {
  it(`should return null on ssr or empty fluid / fixed prop`, () => {
    const emptyImageRef = createPictureRef({})
    expect(emptyImageRef).toBeNull()
  })

  it(`should return null on ssr or empty fluid / fixed prop & wrong type for onLoad`, () => {
    const emptyImageRef = createPictureRef({}, {})
    expect(emptyImageRef).toBeNull()
  })
})

describe(`createPictureRef() with crossOrigin`, () => {
  it(`should set crossOrigin`, () => {
    const emptyImageRef = createPictureRef({
      fluid: fluidShapeMock,
      crossOrigin: `anonymous`,
    })
    expect(emptyImageRef).toMatchInlineSnapshot(`
                                                                                                      <img
                                                                                                        crossorigin="anonymous"
                                                                                                      />
                                                                    `)
  })
})

describe(`createPictureRef() / activatePictureRef() without window`, () => {
  const tmpWindow = global.window
  beforeEach(() => {
    delete global.window
  })
  afterEach(() => {
    global.window = tmpWindow
  })

  it(`should fail without window`, () => {
    const emptyImageRef = createPictureRef({
      fluid: fluidShapeMock,
      critical: true,
    })
    expect(emptyImageRef).toBeNull()
  })

  it(`should fail without window`, () => {
    const testImg = new Image()
    const emptyImageRef = activatePictureRef(testImg, {
      fluid: fluidShapeMock,
      critical: true,
    })
    expect(emptyImageRef).toBeNull()
  })
})

describe(`createPictureRef() with critical image`, () => {
  it(`should preload image on critical`, () => {
    const imageRef = createPictureRef({
      fixed: fixedShapeMock,
      critical: true,
    })
    expect(imageRef).toMatchInlineSnapshot(`
      <img
        src="test_fixed_image.jpg"
        srcset="some srcSet"
      />
    `)
  })
  it(`should preload image on isVisible state`, () => {
    const imageRef = createPictureRef({
      fixed: fixedShapeMock,
      isVisible: true,
    })
    expect(imageRef).toMatchInlineSnapshot(`
      <img
        src="test_fixed_image.jpg"
        srcset="some srcSet"
      />
    `)
  })
  it(`should set empty strings for image on critical without src & srcSet`, () => {
    const fixedMock = { ...fixedShapeMock }
    fixedMock.src = ``
    fixedMock.srcSet = ``
    const emptyImageRef = createPictureRef({
      fixed: fixedMock,
      critical: true,
    })
    expect(emptyImageRef).toMatchInlineSnapshot(`
                                                                                                <img
                                                                                                  src=""
                                                                                                  srcset=""
                                                                                                />
                                                                `)
  })
})

describe(`createPictureRef() with image array`, () => {
  it(`should return imageRef array with fixed Array Mock`, () => {
    const imageRef = createPictureRef({
      ...fixedArrayMock,
      critical: true,
    })
    expect(imageRef).toMatchInlineSnapshot(`
      Array [
        <img
          src="test_fixed_image.jpg"
          srcset="some srcSet"
        />,
        <img
          src="test_fixed_image.jpg"
          srcset="some srcSet"
        />,
      ]
    `)
  })
})

describe(`activatePictureRef() with image array`, () => {
  const dummySelfRef = {
    offsetWidth: 500,
    offsetHeight: 500,
  }
  it(`should return imageRef array with fluid Array Mock`, () => {
    const testImg = new Image()
    const dummyImageRef = activatePictureRef(
      [testImg, testImg],
      {
        ...fluidArrayMock,
        critical: true,
      },
      dummySelfRef
    )
    expect(dummyImageRef).toMatchSnapshot()
  })

  it(`should return imageRef array with fluid Array Mock without selfRef`, () => {
    const testImg = new Image()
    const dummyImageRef = activatePictureRef([testImg, testImg], {
      ...fluidArrayMock,
      critical: true,
    })
    expect(dummyImageRef).toMatchSnapshot()
  })

  it(`should return imageRef array with fixed Array Mock`, () => {
    const testImg = new Image()
    const dummyImageRef = activatePictureRef(
      [testImg, testImg],
      {
        ...fixedArrayMock,
        critical: true,
      },
      dummySelfRef
    )
    expect(dummyImageRef).toMatchSnapshot()
  })
})

describe(`activatePictureRef() with art-direction stack`, () => {
  const dummySelfRef = {
    offsetWidth: 500,
    offsetHeight: 500,
  }
  it(`should return imageRef with fluid art-direction stack`, () => {
    const testImg = new Image()
    const dummyImageRef = activatePictureRef(
      testImg,
      {
        fluid: mockArtDirectionStackFluid,
      },
      dummySelfRef
    )
    expect(dummyImageRef).toMatchSnapshot()
  })
})
