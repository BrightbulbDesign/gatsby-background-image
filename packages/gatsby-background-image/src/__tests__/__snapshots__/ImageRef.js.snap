// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`activatePictureRef() with art-direction stack should return imageRef with fluid art-direction stack 1`] = `
<img
  height="500"
  src="test_fluid_image.jpg"
  srcset="some srcSet"
  width="500"
/>
`;

exports[`activatePictureRef() with image array should return imageRef array with fixed Array Mock 1`] = `
Array [
  <img
    height="500"
    src="test_fixed_image.jpg"
    srcset="some srcSet"
    width="500"
  />,
  <img
    height="500"
    src="test_fixed_image.jpg"
    srcset="some srcSet"
    width="500"
  />,
]
`;

exports[`activatePictureRef() with image array should return imageRef array with fluid Array Mock 1`] = `
Array [
  <img
    height="500"
    src="test_fluid_image.jpg"
    srcset="some srcSet"
    width="500"
  />,
  <img
    height="500"
    src="test_fluid_image.jpg"
    srcset="some srcSet"
    width="500"
  />,
]
`;

exports[`activatePictureRef() with image array should return imageRef array with fluid Array Mock without selfRef 1`] = `
Array [
  <img
    src="test_fluid_image.jpg"
    srcset="some srcSet"
  />,
  <img
    src="test_fluid_image.jpg"
    srcset="some srcSet"
  />,
]
`;
